lib/commands/network.js

```diff
@@ -31,25 +35,51 @@ commands.isWifiOn = async function () {
 commands.setNetworkConnection = async function (type) {
   log.info("Setting network connection");
   // decode the input
-  let airplaneMode = type % 2;
-  type >>= 1;
-  let wifi = type % 2;
-  type >>= 1;
-  let data = type % 2;
+  const shouldEnableAirplaneMode = (type & AIRPLANE_MODE_MASK) !== 0;
+  const shouldEnableWifi = (type & WIFI_MASK) !== 0;
+  const shouldEnableDataConnection = (type & DATA_MASK) !== 0;
 
-  await this.wrapBootstrapDisconnect(async () => {
-    await this.adb.setAirplaneMode(airplaneMode);
-  });
-  await this.wrapBootstrapDisconnect(async () => {
-    await this.adb.broadcastAirplaneMode(airplaneMode);
-  });
-  if (!airplaneMode) {
+  const currentState = await this.getNetworkConnection();
+  const isAirplaneModeEnabled = (currentState & AIRPLANE_MODE_MASK) !== 0;
+  const isWiFiEnabled = (currentState & WIFI_MASK) !== 0;
+  const isDataEnabled = (currentState & DATA_MASK) !== 0;
+
+  if (shouldEnableAirplaneMode !== isAirplaneModeEnabled) {
+    await this.wrapBootstrapDisconnect(async () => {
+      await this.adb.setAirplaneMode(shouldEnableAirplaneMode);
+    });
     await this.wrapBootstrapDisconnect(async () => {
-      await this.setWifiState(wifi);
-      await this.adb.setDataState(data, this.isEmulator());
+      await this.adb.broadcastAirplaneMode(shouldEnableAirplaneMode);
     });
+  } else {
+    log.info(`Not changing airplane mode, since it is already ` +
+             `${shouldEnableAirplaneMode ? 'enabled' : 'disabled'}`);
   }
 
+  if (shouldEnableAirplaneMode && shouldEnableWifi === isWiFiEnabled) {
+    log.info('Not changing data connection, because airplane mode is enabled');
+    return;
+  }
+  if (shouldEnableWifi === isWiFiEnabled && shouldEnableDataConnection === isDataEnabled) {
+    log.info('Not changing data connection/Wi-Fi states, since they are already set to expected values');
+    return;
+  }
+
+  await this.wrapBootstrapDisconnect(async () => {
+    if (shouldEnableWifi !== isWiFiEnabled) {
+      await this.setWifiState(shouldEnableWifi);
+    } else {
+      log.info(`Not changing Wi-Fi state, since it is already ` +
+               `${shouldEnableWifi ? 'enabled' : 'disabled'}`);
+    }
+    if (shouldEnableDataConnection !== isDataEnabled) {
```

Comment:
There is no point to toggle Data connection in Airplane mode.

Reply:
this is handled in the above condition:

```
if (shouldEnableAirplaneMode && shouldEnableWifi === isWiFiEnabled) {
```

Reply:
shouldEnableAirplaneMode = true
shouldEnableWifi = true
shouldEnableDataConnection = true
isWiFiEnabled = false
isDataEnabled = false

Reply:
ok I got it